### PR TITLE
chore: separate Python dependencies for development

### PR DESCRIPTION
**Pull request by bowenliang123**
# Description

- create a Python dependencies requirement for development/testing only, which includes
   - for testing: pytest, mocking and etc.
   - for development: ruff linter and etc.
- help to reduce docker image size for Python dependencies

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] TODO

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

File: api/requirements-dev.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+coverage~=7.2.4
+pytest~=7.3.1
+pytest-mock~=3.11.1
+pytest-benchmark~=4.0.0
+ruff~=0.3.5
```
Comment (LeoQuote) [inline api/requirements-dev.txt]:
```suggestion
ruff~=0.3.5
-r requirements.txt
```

or on the top

Comment (bowenliang123) [reply api/requirements-dev.txt]: Hi, the `requirements.txt` is for production dependencies, which is not included in `requirements-dev.txt` on purpose by design. I'm trying to keep `requirements-dev.txt` small and available for testing, lining and other development purposes only.

File: .github/workflows/style.yml [hunk 2]
```diff
@@ -22,9 +22,11 @@ jobs:
         uses: actions/setup-python@v5
         with:
           python-version: '3.10'
+          cache: 'pip'
+          cache-dependency-path: ./api/requirements-dev.txt
 
       - name: Python dependencies
-        run: pip install ruff
+        run: pip install -r ./api/requirements-dev.txt
```
Comment (takatost) [inline .github/workflows/style.yml]: I think you only need to install ruff here.

File: dev/reformat [hunk 3]
```diff
@@ -5,7 +5,7 @@ set -x
 # python style checks rely on `ruff` in path
 if ! command -v ruff &> /dev/null; then
     echo "Installing Ruff ..."
-    pip install ruff
+    pip install -r ./api/requirements-dev.txt
```
Comment (takatost) [inline dev/reformat]: I think you only need to install ruff here.

**Review by takatost — CHANGES_REQUESTED** [review]

File: .github/workflows/style.yml (hunk 2 above)
Comment (bowenliang123) [reply .github/workflows/style.yml]: The general benefits for putting `ruff` in  `requirements-dev.txt` here include:
1. pin the required packages including linters in `requirements-dev.txt` as well
2. `requirements-dev.txt` is designed for general development purpose, covering testing, lining and etc.
3. `requirements-dev.txt` helps to cache the pip packages in Github actions with same hash for testing and lining, avoiding repeated downloading and saving execution time and cost.

Comment (takatost) [reply .github/workflows/style.yml]: I think the style check task here only relies on the ruff package, the others are redundant. And I believe that only installing the ruff package and using action cache doesn't make much of a difference in terms of time consumption. 

P.S. I think we should follow the Pragmatic Programming principle, avoid blindly following established norms, and hope that each piece of code has its most reasonable explanation in decision-making.

Comment (bowenliang123) [reply .github/workflows/style.yml:29]: No problem. Will fix it and follow your suggestions.

Comment (bowenliang123) [reply .github/workflows/style.yml]: Have fixed the problem, by following your advice.

**Comment by takatost** [comment]
Could you please separately explain in the README about the requirements-dev.txt for running tests?

**Comment by bowenliang123** [comment]
I shall create a PR soon bringing a new script for running tests on both local env and Github actions. How about updating the description of `requirements-dev.txt` and the script in README at that time?

**Review by takatost — APPROVED** [review]
LGTM

**Pull request merged**